Immersive dark mode + Theme Listener

WilliamABradley opened the pull request:
## Summary of the Pull Request

**What is this about:**

Handles issue #13712 to add dark mode titlebars.

![Screenshot (23)](https://user-images.githubusercontent.com/6026537/168509316-5da1347c-d15a-46e3-b6e3-fa8cbd0a5ec1.png)

![Screenshot (24)](https://user-images.githubusercontent.com/6026537/168814252-220e9e84-c5ff-44e1-8bb8-e1e5fd94e413.png)

**What is included in the PR:** 

**How does someone test / validate:** 

## Quality Checklist

- [x] **Linked issue:** #13712 
- [x] **Communication:** I've discussed this with core contributors in the issue. 
- [x] **Tests:** Added/updated and all pass
- [x] **Installer:** Added/updated and all pass
- [x] **Localization:** All end user facing strings can be localized
- [x] **Docs:** Added/ updated
- [x] **Binaries:** Any new files are added to WXS / YML
   - [x] No new binaries
   - [x] [YML for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/pipeline.user.windows.yml#L68) for new binaries
   - [x] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries

## Contributor License Agreement (CLA)
A CLA must be signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA.

Jay-o-Way commented:
CC @niels9001 because he is working on UI "overhauling" too.
@WilliamABradley Please incluse a screenshot :)

WilliamABradley commented:
I added the screenshot 🙂

Once the logic is ready for PowerRename, should I make separate PRs for the other apps or make it an all in one?

WilliamABradley commented:
@stefansjfw @michael-hawker @niels9001 @Jay-o-Way I conjured up some real magic here 😄 

Since WinUI 3 deprecated UISettings, I had to implement a new theme listener implemented via the Registry and `RegNotifyChangeKeyValue`.

I scoured the internet and this seemed to be the simplest solution without rewriting the entrypoint to access the message pump, which feels gross? (Although this requires a sacrifice of 1 thread, although the callback is marked asynchronous, so unsure if this is a busy wait or a sleep action.

Unless I'm wrong, that would be the only way to receive those raw messages right?

Anyway, this is my first time in C++ so go easy on me, and let me know if I'm doing anything wrong 🙂 

Once this code is validated, I will make a new shared unmanaged library for it, and then I think we can use it in all of the toys?

WilliamABradley commented:
The magic in action!
It is very responsive!

https://user-images.githubusercontent.com/6026537/168764112-077772fe-0a04-4c63-80dd-723d384551c9.mp4

WilliamABradley commented:
I moved the C++ Theming code into the Themes library as that seemed where appropriate code was.
I also replaced the logic of one function call that checked for dark mode, which depended on UISettings, which is deprecated in WinUI3.

I also ported the C++ ThemeHelpers code to .NET and applied it to Settings and OOBE:

![Screenshot (24)](https://user-images.githubusercontent.com/6026537/168814252-220e9e84-c5ff-44e1-8bb8-e1e5fd94e413.png)

Eldenroot commented:
Do not forget to change system tray menu and context menu too

WilliamABradley commented:
@Eldenroot That is a different issue that can't be solved with the Dwm Attributes I don't think

WilliamABradley commented:
Update @niels9001, I believe this is ready for review.

I wasn't too sure about Unit Tests, since these changes are intrinsic to the environment.
I have tested thoroughly on my machine through all of the different potential states, so I'm wondering what the next step is?

niels9001 commented:
@WilliamABradley Awesome! I wonder if we should extend the content into the titlebar in this PR.. or leave it for another one. WinUi has APIs for this but not sure if those will work downlevel to W10, so might be better to hold out for those.👍

WilliamABradley commented:
I think that would be a topic for another PR 🙂

Aaron-Junker reviewed (approved):
Tested and works as expected.

But there's a small delay when changing light/dark mode option from the changing title bar to the changing content.

WilliamABradley commented:
Yeah, I'm not sure we can do much about that, hopefully Windows App Sdk will improve the situation in the future with a native API, or at least inprove the ExtendIntoTitlebar support.

Right now I don't think we will be able to get real time transitions, but Windows doesn't have day/night cycling like MacOS, so the user would only see that when changing themes.

stefansjfw reviewed (changes requested):
`runner` needs `dwmapi.lib` as AdditionalDependency to fix build failure

stefansjfw commented on src/modules/powerrename/PowerRenameUILib/PowerRenameUI.vcxproj:
```diff
@@ -68,6 +68,9 @@
       <PreprocessorDefinitions>_DEBUG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
     </ClCompile>
+    <Link>
+      <AdditionalDependencies Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">kernel32.lib;user32.lib;dwmapi.lib;%(AdditionalDependencies)</AdditionalDependencies>
```
this should be unconditional

WilliamABradley replied:
Fixed 🙂

stefansjfw commented on src/common/Themes/Themes.vcxproj:
```diff
@@ -31,10 +31,14 @@
   </ItemGroup>
   <ItemGroup>
     <ClCompile Include="icon_helpers.cpp" />
+    <ClCompile Include="ThemeListener.cpp" />
```
Why different naming styles ? :)

WilliamABradley replied:
I created a class, while the others were loose c++ helper files, and this library had that convention.
Should I change the class naming style, or keep the class naming style, but change it's file name?

I'm new to c++, so I'm not really sure what the convention is 😅

stefansjfw replied:
coding style is ok. Just file naming should be uniform, i.e. rename `ThemeListener.h|cpp` -> `theme_listener.h|cpp`

stefansjfw commented on src/common/ManagedCommon/ManagedCommon.csproj:
```diff
@@ -17,6 +17,8 @@
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
+    <PackageReference Include="Microsoft.Win32.Registry" Version="5.0.0" />
+    <PackageReference Include="System.Management" Version="6.0.0" />
```
this requires installer update

stefansjfw replied:
Check Product.wxs file. You need to ship Microsoft.Win32.Registry and  System.Management dlls with binaries that are using those dlls

WilliamABradley replied:
@stefansjfw I updated the Product.wxs file, can you confirm if I did it correctly?

WilliamABradley commented:
@stefansjfw I made the changes to remove the PowerRename deps condition, and add dwmapi.lib to the runner.
Let me know what I need to do for your other comments, if anything 🙂

WilliamABradley commented:
@stefansjfw I'm not quite sure what I am doing with the product.wxs file 😅 
Could you let me know what I'm doing wrong, or if there is someone/somewhere else I should get help from?

stefansjfw commented:
> @stefansjfw I'm not quite sure what I am doing with the product.wxs file 😅 Could you let me know what I'm doing wrong, or if there is someone/somewhere else I should get help from?

Looking into it, will let you know soon :)

stefansjfw commented on src/common/Themes/theme_helpers.cpp:
```diff
@@ -0,0 +1,111 @@
+        return info;
+    }
+
+    DWORD WINAPI CheckImmersiveDarkMode(LPVOID lpParam)
```
where is this used?

WilliamABradley replied:
Whoops! This should have been removed.
Good catch!

stefansjfw reviewed (commented):
Code looks good, and works good as well. For product wxs, just remove `Microsoft.Win32.Registry.dll` everywhere and this should be good to go

WilliamABradley commented:
@stefansjfw made the changes 🙂

stefansjfw commented on src/common/Themes/theme_listener.cpp:
```diff
@@ -0,0 +1,57 @@
+#include "ThemeListener.h"
```
Build fails because of this

WilliamABradley replied:
🤦

stefansjfw commented:
Very nice work! Thanks for the contribution!

crutkas commented:
Should this be in .59?

crutkas commented:
Quickly looking at this, it feels like we have duplicate implications. Can’t the c# call the c++ if needed?

WilliamABradley commented:
@crutkas the binding code for the C# to the C++ layer would be more than the actual C# implementation.
Also, in managed code, the C# implementation should be better, and feel more native to C#.

WilliamABradley commented:
@crutkas @stefansjfw is there anything else required to get this merged?

crutkas commented:
IMO, I would want the c# to interoperability to the c++ versus the same functionality in two spots. This will lead to one being updated and not the other. 

@jaimecbernardo / @DHowett thoughts?

DHowett commented:
> IMO, I would want the c# to interoperability to the c++ versus the same functionality in two spots. This will lead to one being updated and not the other.
> 
> @jaimecbernardo / @DHowett thoughts?

Thoughts: it's just code, we can change it in the future if it's a problem.

You know the rule of three! If you copy/paste the same code three times, _then_ you should consider making it a helper.

DHowett commented on src/common/ManagedCommon/ThemeHelpers.cs:
```diff
@@ -0,0 +1,65 @@
+
+            switch (theme)
+            {
+                case "hc1":
+                    return Theme.HighContrastOne;
+                case "hc2":
+                    return Theme.HighContrastTwo;
+                case "hcwhite":
+                    return Theme.HighContrastWhite;
+                case "hcblack":
+                    return Theme.HighContrastBlack;
```
I don't believe this is the accepted way to determine whether HC mode is enabled.

Terminal uses this (approved by the window manager folks, same method XAML uses)

https://github.com/microsoft/terminal/blob/ed27737233714dea77877624d1beeb49e2ccd36e/src/cascadia/WindowsTerminal/icon.cpp#L14-L25

WilliamABradley replied:
I took this from another piece of PowerToys, but since I don't actually use it, I think I will just remove it.

WilliamABradley replied:
Removed!

DHowett commented on src/common/ManagedCommon/ThemeHelpers.cs:
```diff
@@ -0,0 +1,65 @@
+
+        public static bool SupportsImmersiveDarkMode()
+        {
+            return Environment.OSVersion.Version.Major >= 10 && Environment.OSVersion.Version.Build >= 18985;
```
It is generally better to simply try to call the API and see if it fails. If it fails, don't pursue it. :smile:

WilliamABradley replied:
You reckon that set an attribute that doesn't exist in previous versions is fine?
What would be the result of the undefined behaviour?

WilliamABradley replied:
I tested it with a fake DWM Attribute (40) and it doesn't do anything!
Removed all of this logic!

DHowett replied:
Thanks! Yeah, DWM attributes are interesting. Apart from a few internal reserved ones, the highest number you see in the documentation is the "final" attribute. That means that using a higher number on an older version of Windows is a no-op: it doesn't know about it, so it won't do anything!

DHowett commented on src/common/Themes/theme_helpers.cpp:
```diff
@@ -0,0 +1,79 @@
+
+namespace
+{
+    RTL_OSVERSIONINFOW GetRealOSVersion()
```
Prefer VerifyVersionInfoW and _make sure your app manifest says you run on Windows 10_. There's no need to use internal ntdll APIs.

DHowett replied:
But again, earlier comment about just calling the API and seeing if it fails. No big deal if it does, it won't break anything!

WilliamABradley replied:
Removed all of this logic!

WilliamABradley commented:
@DHowett Made the changes 🙂

crutkas commented:
So we already have https://github.com/microsoft/PowerToys/blob/main/src/common/Common.UI/ThemeManager.cs

why are we duplicating stuff?

WilliamABradley commented:
> So we already have https://github.com/microsoft/PowerToys/blob/main/src/common/Common.UI/ThemeManager.cs
> 
> why are we duplicating stuff?

This handles swapping out the Theme Dictionaries of the app, not system theme, they do different things.

WilliamABradley commented:
ThemeManager is also only used for changing the theme dictionary of the WPF library used in only some toys, and not the Settings App.

Settings App does not reference that Common.UI library, and we wouldn't want to bring in a WPF dependency on a WinUI app right?

WilliamABradley commented:
@crutkas @DHowett @stefansjfw is there anything preventing this from being merged?

WilliamABradley commented:
@crutkas @DHowett @stefansjfw is there anything preventing this from being merged?

stefansjfw commented:
> @crutkas @DHowett @stefansjfw is there anything preventing this from being merged?

not from my side

DHowett commented:
Not to my knowledge from my side! Sorry for the delay.

WilliamABradley commented:
@stefansjfw resolved conflicts 🙂

jaimecbernardo commented:
Thanks a lot for the contribution!
